### PR TITLE
fix: convert measurements to indices without allocating a high-dimens…

### PR DESCRIPTION
**Pull request by ajberdy**
…ional array

*Issue #, if available:*

Probability result type in the SDK allocates a numpy array of size 2 ^ num_qubits which fails with the error :
ValueError: maximum supported dimension for an ndarray is 32, found 40

This is an issue for SV1 which goes up to 34 qubits. 

*Description of changes:*

Instead of using ravel_multi_index, which allocates an ndarray with n dimensions (n qubits), multiply by a vector of powers of two to get the index in a single matrix multiplication.

*Testing done:*

it works now

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md) doc
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-sdk-python/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-sdk-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by speller26 — DISMISSED** [review]
*Much* more understandable to boot

File: test/unit_tests/braket/tasks/test_gate_model_quantum_task_result.py [hunk 1]
```diff
@@ -420,3 +420,41 @@ def test_hash_result_types(observable_1, observable_2):
...
+            "{'type': <Type.statevector: 'statevector'>}",
+        ),
+        (
+            jaqcd.Variance(observable=["y", "z"], targets=[2, 1]),
+            "{'observable': ['y', 'z'], 'targets': [2, 1], 'type': <Type.variance: 'variance'>}",
+        ),
+    ],
+)
+def test_result_type_hash(result_type, rt_hash):
+    assert GateModelQuantumTaskResult._result_type_hash(result_type) == rt_hash
```
Comment (speller26) [inline test/unit_tests/braket/tasks/test_gate_model_quantum_task_result.py:460]: The uncovered lines are actually different: https://github.com/aws/amazon-braket-sdk-python/blob/main/src/braket/tasks/gate_model_quantum_task_result.py#L35

Comment (speller26) [reply test/unit_tests/braket/tasks/test_gate_model_quantum_task_result.py:460]: In fact, I think that function isn't used at all; you can then remove this test altogether, as, it looks like the correct method already has coverage, and we don't want to explicitly test private methods

Comment (ajberdy) [reply test/unit_tests/braket/tasks/test_gate_model_quantum_task_result.py:460]: yup that has no usages and should've been deleted

**Review by speller26 — APPROVED** [review]

**Pull request merged**